apps/fipsinstall.c

```diff
@@ -0,0 +1,430 @@
+    if (vtmp != NULL) {
+        *vtmp = 0;
+        vtmp++;
+    }
+    rv = EVP_MAC_ctrl_str(ctx, stmp, vtmp);
+    OPENSSL_free(stmp);
+    return rv;
+}
+
+static int do_mac(EVP_MAC_CTX *ctx, unsigned char *tmp, BIO *in,
+                  unsigned char *out, size_t *out_len)
+{
+    int ret = 0;
+    int i;
+
+    if (!EVP_MAC_init(ctx))
+        goto err;
+    if (EVP_MAC_size(ctx) > *out_len)
+        goto end;
+    for (;;) {
+        i = BIO_read(in, (char *)tmp, BUFSIZE);
+        if (i < 0)
+            goto err;
+        if (i == 0)
+            break;
+        if (!EVP_MAC_update(ctx, tmp, i))
+            goto err;
+    }
+end:
+    if (!EVP_MAC_final(ctx, out, out_len))
+        goto err;
+    ret = 1;
+err:
+    return ret;
+}
+
+static int load_fips_prov_and_run_self_test(const char *prov_name)
+{
+    int ret = 0;
+    OSSL_PROVIDER *prov = NULL;
+
+    prov = OSSL_PROVIDER_load(NULL, prov_name);
+    if (prov == NULL) {
+        BIO_printf(bio_err, "Failed to load FIPS module\n");
+        goto end;
+    }
+    ret = 1;
+end:
+    OSSL_PROVIDER_unload(prov);
+    return ret;
+}
+
+
+static int print_mac(BIO *bio, const char *label, const unsigned char *mac,
+                     size_t len)
+{
+    int ret;
+    char *hexstr = NULL;
+
+    hexstr = OPENSSL_buf2hexstr(mac, (long)len);
+    if (hexstr == NULL)
+        return 0;
+    ret = BIO_printf(bio, "%s = %s\n", label, hexstr);
+    OPENSSL_free(hexstr);
+    return ret;
+}
+
+static int write_config_header(BIO *out, const char *prov_name,
+                               const char *section)
+{
+    return BIO_printf(out, "openssl_conf = openssl_init\n\n")
+           && BIO_printf(out, "[openssl_init]\n")
+           && BIO_printf(out, "%s = %s\n\n", prov_name, section);
+}
+
+/*
+ * Outputs a fips related config file that contains entries for the fips
+ * module checksum and the installation indicator checksum.
+ *
+ * Returns 1 if the config file is written otherwise it returns 0 on error.
+ */
+static int write_config_fips_section(BIO *out, const char *section,
+                                     unsigned char *module_mac,
+                                     size_t module_mac_len,
+                                     unsigned char *install_mac,
+                                     size_t install_mac_len)
+{
+    int ret = 0;
+
+    if (!(BIO_printf(out, "[%s]\n", section) > 0
+          && BIO_printf(out, "%s = %s\n", OSSL_PROV_PARAM_INSTALL_VERSION,
+                        VERSION_VAL) > 0
+          && print_mac(out, OSSL_PROV_PARAM_MODULE_MAC, module_mac,
+                       module_mac_len)))
+        goto end;
+
+    if (install_mac != NULL) {
+        if (!(print_mac(out, OSSL_PROV_PARAM_INSTALL_MAC, install_mac,
+                      install_mac_len)
+              && BIO_printf(out, "%s = %s\n", OSSL_PROV_PARAM_INSTALL_STATUS,
+                          INSTALL_STATUS_VAL) > 0))
+        goto end;
+    }
+    ret = 1;
+end:
+    return ret;
+}
```

Comment:
If the module isn't in a standard location or the location given with `$OPENSSL_MODULES`, you will also need this line in that section:

```
module = {module_fname}
```

Reply:
Ok so this 'module' parameter is not really optional at the moment since I need to calculate a mac on the module file (so I need its path and extension - in order to  digest the entire file).
I dont think using the DSO convert function and taking on the extension is a really good way to do this.

Reply:
If the module path isn't given, you could construct it from known facts.  What you need is a location (`$OPENSSL_MODULES` with the default modules directory as fallback (*)), the file name itself (`prov_name`) and the extension (which is platform dependent, but quite easy to get right)

(*) #8709 is a PR to help applications get exactly that information

Reply:
Requiring the module pathname doesn't seem unreasonable; the alternative is having to understand what is going on with openssl runtime config options.

Reply:
I'm with @richsalz 
We shouldn't be constructing the module path inside the module.  Pass it in unconditionally.